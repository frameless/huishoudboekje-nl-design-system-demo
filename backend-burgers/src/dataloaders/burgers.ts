import fetch from "node-fetch";
import {createServiceUrl} from "../config/services";

const BurgerLoader = {

	findAll: async () => {
		return await fetch(createServiceUrl("burgers", "/burgers")).then(r => r.json()).then(r => r.data);
	},

	findById: async (id: number) => {
		return await fetch(createServiceUrl("burgers", `/burgers/${id}`)).then(r => r.json()).then(r => r.data);
	},

};

export default BurgerLoader;