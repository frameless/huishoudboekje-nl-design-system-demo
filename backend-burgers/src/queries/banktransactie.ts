import {intArg, nonNull} from "nexus";
import {Context} from "../context";
import {isDev} from "../utils/things";

const banktransactie = (t) => {
	t.field("banktransactie", {
		args: {
			id: nonNull(intArg()),
		},
		type: "Banktransactie",
		resolve: async (_, args, ctx: Context) => {
			if (!args.id) {
				return null;
			}

			return await ctx.dataSources.banktransactieservice.getBanktransactieById(args.id)
		},
	});

	if (!isDev) {
		return;
	}

	/* Fields below here are only available when an envvar NODE_ENV="development". */
	// t.list.field("banktransacties", {
	// 	type: Banktransactie,
	// 	resolve: async (root, args) => {
	// 		return await DataLoader.getAllBanktransacties();
	// 	},
	// });
};

export default banktransactie;
