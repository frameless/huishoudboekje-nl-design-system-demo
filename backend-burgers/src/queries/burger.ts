import {intArg} from "nexus";
import {Context} from "../context";
import Burger from "../models/Burger";
import {isDev} from "../utils/things";

const burger = (t) => {
	t.field("burger", {
		args: {
			bsn: intArg(),
		},
		type: Burger,
		resolve: async (_, args, ctx: Context) => {
			const burgers = await ctx.dataSources.huishoudboekjeservice.getBurgersByBsns([args.bsn]);
			return burgers.shift();
		},
	});

	if (!isDev) {
		return;
	}

	/* Fields below here are only available when an envvar NODE_ENV="development". */

	t.list.field("burgers", {
		type: "Burger",
		resolve: (_, __, ctx: Context) => {
			return ctx.dataSources.huishoudboekjeservice.getAllBurgers();
		},
	});

};

export default burger;
