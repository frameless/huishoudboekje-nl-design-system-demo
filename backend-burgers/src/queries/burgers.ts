import BurgerLoader from "../dataloaders/burgers";
import Burger from "../models/Burger";

const burgers = (t) => {
	t.list.field("burgers", {
		type: Burger,
		resolve: (root, args, ctx) => {
			return BurgerLoader.findAll();
		},
	});
};

export default burgers;
