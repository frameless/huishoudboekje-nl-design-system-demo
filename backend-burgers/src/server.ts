import {ApolloServer} from "apollo-server";
import {ApolloServerPluginLandingPageGraphQLPlayground} from "apollo-server-core/dist/plugin/landingPage/graphqlPlayground";
import dotenv from "dotenv";
import Banktransactieservice from "./datasources/banktransactieservice";
import Huishoudboekjeservice from "./datasources/huishoudboekjeservice";
import Organisatieservice from "./datasources/organisatieservice";
import schema from "./schema";
import debugPlugin from "./utils/debugPlugin";
import {isDev} from "./utils/things";

dotenv.config();

const server = new ApolloServer({
	schema,
	cors: {
		credentials: true,
	},
	dataSources: () => ({
		huishoudboekjeservice: new Huishoudboekjeservice(),
		organisatieservice: new Organisatieservice(),
		banktransactieservice: new Banktransactieservice(),
	}),
	debug: isDev,
	plugins: [
		ApolloServerPluginLandingPageGraphQLPlayground(),
		...(process.env.NODE_ENV !== "test") ? [debugPlugin()] : [],
	],
});

export default server;
