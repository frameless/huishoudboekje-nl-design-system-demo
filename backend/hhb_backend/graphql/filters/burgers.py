import graphene

from ..scalars.complex_filter import ComplexFilterType


class BurgerFilter(graphene.InputObjectType):
    OR = graphene.Field(lambda: BurgerFilter)
    AND = graphene.Field(lambda: BurgerFilter)
    id = ComplexFilterType()
    telefoonnummer = ComplexFilterType()
    email = ComplexFilterType()
    geboortedatum = ComplexFilterType()
    iban = ComplexFilterType()
    achternaam = ComplexFilterType()
    huisnummer = ComplexFilterType()
    postcode = ComplexFilterType()
    straatnaam = ComplexFilterType()
    voorletters = ComplexFilterType()
    voornamen = ComplexFilterType()
    plaatsnaam = ComplexFilterType()
    huishouden = ComplexFilterType()
