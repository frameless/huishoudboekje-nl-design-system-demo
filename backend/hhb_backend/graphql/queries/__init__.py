""" GraphQL schema queries module """
import graphene
from .gebruikers import GebruikersQuery, GebruikerQuery
from .organisaties import OrganisatieQuery, OrganisatiesQuery
from .afspraken import AfspraakQuery, AfsprakenQuery

class RootQuery(graphene.ObjectType):
    """ The root of all queries """
    gebruiker = GebruikerQuery.return_type
    gebruikers = GebruikersQuery.return_type
    organisatie = OrganisatieQuery.return_type
    organisaties = OrganisatiesQuery.return_type
    afspraak = AfspraakQuery.return_type
    afspraken = AfsprakenQuery.return_type

    def resolve_gebruiker(root, info, **kwargs):
        return GebruikerQuery.resolver(root, info, **kwargs)

    def resolve_gebruikers(root, info, **kwargs):
        return GebruikersQuery.resolver(root, info, **kwargs)

    def resolve_organisatie(root, info, **kwargs):
        return OrganisatieQuery.resolver(root, info, **kwargs)

    def resolve_organisaties(root, info, **kwargs):
        return OrganisatiesQuery.resolver(root, info, **kwargs)

    def resolve_afspraak(root, info, **kwargs):
        return AfspraakQuery.resolver(root, info, **kwargs)

    def resolve_afspraken(root, info, **kwargs):
        return AfsprakenQuery.resolver(root, info, **kwargs)
