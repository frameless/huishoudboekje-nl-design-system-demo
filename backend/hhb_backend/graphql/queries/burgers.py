""" GraphQL Burgers query """
import logging
import graphene

import hhb_backend.graphql.models.burger as burger
from hhb_backend.audit_logging import AuditLogging
from hhb_backend.graphql.dataloaders import hhb_dataloader
from hhb_backend.graphql.scalars.dynamic_types import DynamicType
from hhb_backend.graphql.utils.dates import valid_afspraak
from hhb_backend.graphql.utils.gebruikersactiviteiten import GebruikersActiviteitEntity


class BurgerQuery:
    return_type = graphene.Field(burger.Burger, id=graphene.Int(required=True))

    @classmethod
    def resolver(cls, _, info, id):
        logging.info(f"Get burger")
        result = hhb_dataloader().burgers.load_one(id)
        AuditLogging.create(
            action=info.field_name,
            entities=[
                GebruikersActiviteitEntity(entityType="burger", entityId=id)
            ],
        )
        return result


class BurgersQuery:
    return_type = graphene.List(
        burger.Burger,
        ids=graphene.List(graphene.Int),
        search=DynamicType()
    )

    @classmethod
    def resolver(cls, _, info, **kwargs):
        logging.info(f"Get burgers")
        if "ids" in kwargs:
            burgers = hhb_dataloader().burgers.load(kwargs["ids"])
            AuditLogging.create(
                action=info.field_name,
                entities=[
                    GebruikersActiviteitEntity(entityType="burger", entityId=id)
                    for id in kwargs["ids"]
                ]
            )
            return burgers

        if "search" in kwargs:
            search = str(kwargs["search"]).lower()
            if search:            
                burger_ids = set()
                afspraken_ids = set()

                burgers = hhb_dataloader().burgers.load_all(filters=kwargs.get("filters", None))
                for burger in burgers:
                    if search in str(burger['achternaam']).lower() or \
                        search in str(burger['voornamen']).lower() or \
                        search in str(burger['bsn']).lower():
                        burger_ids.add(burger["id"])

                rekeningen = hhb_dataloader().rekeningen.load_all(filters=kwargs.get("filters", None))
                for rekening in rekeningen:
                    if search in str(rekening['iban']).lower() or \
                        search in str(rekening['rekeninghouder']).lower():
                        for burger_id in rekening["burgers"]:
                            if burger_id:
                                burger_ids.add(burger_id)
                        for afspraak_id in rekening["afspraken"]:
                            afspraken_ids.add(afspraak_id)

                afspraken = hhb_dataloader().afspraken.load_all(filters=kwargs.get("filters", None))
                for afspraak in afspraken:
                    if valid_afspraak(afspraak) and afspraak["burger_id"]:
                        if afspraak["id"] in afspraken_ids or search in str(afspraak['zoektermen']).lower():
                            burger_ids.add(afspraak["burger_id"])

                result = []
                for burger in burgers:
                    if burger["id"] in burger_ids:
                        result.append(burger)

                AuditLogging.create(
                    action=info.field_name,
                    entities=[
                        GebruikersActiviteitEntity(entityType="burger", entityId=burger["id"])
                        for burger in result
                    ]
                )
                return result

        burgers = hhb_dataloader().burgers.load_all(filters=kwargs.get("filters", None))
        AuditLogging.create(
            action=info.field_name,
            entities=[
                GebruikersActiviteitEntity(entityType="burger", entityId=burger.id)
                for burger in burgers
            ] if "filters" in kwargs else []
        )
        return burgers


class BurgersPagedQuery:
    return_type = graphene.Field(
        burger.BurgersPaged,
        start=graphene.Int(),
        limit=graphene.Int()
    )

    @classmethod
    def resolver(cls, _, info, **kwargs):
        logging.info(f"Get burgers paged")
        if "start" in kwargs and "limit" in kwargs:
            burgers = hhb_dataloader().burgers.load_paged(start=kwargs["start"], limit=kwargs["limit"])
        else:
            burgers = hhb_dataloader().burgers.load_all()

        AuditLogging.create(
            action=info.field_name,
            entities=[
                GebruikersActiviteitEntity(entityType="burger", entityId=burger["id"])
                for burger in burgers["burgers"]
            ] if "start" in kwargs and "limit" in kwargs else []
        )

        return burgers
