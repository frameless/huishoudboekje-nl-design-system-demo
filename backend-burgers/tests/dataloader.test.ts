import Huishoudboekjeservice from "../src/datasources/huishoudboekjeservice";
import server from "../src/server";
import burgers from "./fixtures/burgers.json";
import {GetOneBurger} from "./queries";

jest.mock("../src/datasources/huishoudboekjeservice");

describe("BurgerLoader", () => {

	describe("Given the burger exists", () => {

		it("gets the entire burger", async () => {
			jest.spyOn(Huishoudboekjeservice.prototype, "getBurgersByBsns").mockImplementation(async (bsns) => burgers.filter(b => bsns.includes(b.bsn)));

			const result = await server.executeOperation({
				query: GetOneBurger,
				variables: {
					bsn: burgers[0].bsn,
				},
			});

			expect(result.errors).toEqual(undefined);
			expect(result.data).toMatchSnapshot();
		});

	});

});