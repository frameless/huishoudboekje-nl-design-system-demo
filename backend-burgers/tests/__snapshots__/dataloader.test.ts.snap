// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BurgerLoader Given the burger exists gets the entire burger 1`] = `
Object {
  "burger": Object {
    "achternaam": "de Jager",
    "afspraken": null,
    "bsn": "999999990",
    "email": "fien.de.jager@sloothuizen.nl",
    "geboortedatum": "1987-06-18",
    "huisnummer": "14",
    "id": 1,
    "plaatsnaam": "Sloothuizen",
    "postcode": "9999ZZ",
    "rekeningen": null,
    "straatnaam": "Winkelstraat",
    "telefoonnummer": "0687542132",
    "voorletters": "F.S.",
    "voornamen": "Fien Sandra",
  },
}
`;
